tokio-timer/src/timer/mod.rs

```diff
@@ -0,0 +1,538 @@
+/// intervals of time. At each level, the wheel has 64 slots. Each slot covers a
+/// range of time equal to the wheel at the lower level. At level zero, each
+/// slot represents one millisecond of time.
+///
+/// The wheels are:
+///
+/// * Level 0: 64 x 1 millisecond slots.
+/// * Level 1: 64 x 64 millisecond slots.
+/// * Level 2: 64 x ~4 second slots.
+/// * Level 3: 64 x ~4 minute slots.
+/// * Level 4: 64 x ~4 hour slots.
+/// * Level 5: 64 x ~12 day slots.
+///
+/// When the timer processes entries at level zero, it will notify all the
+/// [`Sleep`] instances as their deadlines have been reached. For all higher
+/// levels, all entries will be redistributed across the wheel at the next level
+/// down. Eventually, as time progresses, entries will `Sleep` instances will
+/// either be canceled (dropped) or their associated entries will reach level
+/// zero and be notified.
+///
+/// [`Sleep`]: ../struct.Sleep.html
+/// [`Interval`]: ../struct.Interval.html
+/// [`Deadline`]: ../struct.Deadline.html
+/// [paper]: http://www.cs.columbia.edu/~nahum/w6998/papers/ton97-timing-wheels.pdf
+#[derive(Debug)]
+pub struct Timer<T, N = SystemNow> {
+    /// Shared state
+    inner: Arc<Inner>,
+
+    /// The number of milliseconds elapsed since the timer started.
+    elapsed: u64,
+
+    /// Timer wheel.
+    ///
+    /// Levels:
+    ///
+    /// * 1 ms slots / 64 ms range
+    /// * 64 ms slots / ~ 4 sec range
+    /// * ~ 4 sec slots / ~ 4 min range
+    /// * ~ 4 min slots / ~ 4 hr range
+    /// * ~ 4 hr slots / ~ 12 day range
+    /// * ~ 12 day slots / ~ 2 yr range
+    levels: Vec<Level>,
+
+    /// Thread parker. The `Timer` park implementation delegates to this.
+    park: T,
+
+    /// Source of "now" instances
+    now: N,
+}
+
+/// Return value from the `turn` method on `Timer`.
+///
+/// Currently this value doesn't actually provide any functionality, but it may
+/// in the future give insight into what happened during `turn`.
+#[derive(Debug)]
+pub struct Turn(());
+
+pub(crate) struct Inner {
+    /// The instant at which the timer started running.
+    start: Instant,
+
+    /// The last published timer `elapsed` value.
+    elapsed: AtomicU64,
+
+    /// Number of active timeouts
+    num: AtomicUsize,
+
+    /// Head of the "process" linked list.
+    process: entry::AtomicStack,
+
+    /// Unparks the timer thread.
+    unpark: Box<Unpark>,
+}
+
+/// Number of levels
+const NUM_LEVELS: usize = 6;
+
+/// The maximum duration of a sleep
+const MAX_DURATION: u64 = 1 << (6 * NUM_LEVELS);
+
+/// Maximum number of timeouts the system can handle concurrently.
+const MAX_TIMEOUTS: usize = usize::MAX >> 1;
+
+// ===== impl Timer =====
+
+impl<T> Timer<T>
+where T: Park
+{
+    pub fn new(park: T) -> Self {
+        Timer::new_with_now(park, SystemNow::new())
+    }
+}
+
+impl<T, N> Timer<T, N> {
+    /// Returns a reference to the underlying `Park` instance.
+    pub fn get_ref(&self) -> &T {
```

Comment:
I wouldn't immediately assume `Timer::get_ref()` returns the `Park`. Maybe `get_park()` and `get_park_mut()`?

Reply:
Ok, I hesitated between both options, so I can switch to `get_park`.